feat(conductor): handle case where block is received from DA before gossip

noot opened the pull request:
as title says, if we receive a block from DA that we don't have an execution hash for (ie it was not executed either because it was never received, or there was nothing to execute) we try to execute it, and if there was a block executed, we mark it final. "nothing to execute" case means there were no txs for our rollup namespace in the block.

closes #13

SuperFluffy commented on crates/astria-conductor/src/executor.rs:
```diff
@@ -238,38 +238,62 @@ impl Executor {
         Ok(())
     }
+
+    async fn finalize_block(
```
Let's start instrumenting all new methods we are writing, in this case:

```rust
#[instrument(skip_all, fields(execution_block_hash = hex::encode(execution_block_hash), %sequencer_block_hash))]```

(Not sure about `sequencer_block_hash`; I think it implemented `Display`, so I put in the `%` sigil).

SuperFluffy replied:
We totally should make a push for this. But at the same time it would also be ok to slowly "instrumentify" things for each function we touch IMO

SuperFluffy reviewed (changes requested):
The logic looks fine to me. Still requesting some changes so we don't overlook adding doc comments, instrumentation and wrapping errors.

Have you thought about we can test this logic? We should start mocking the interfaces we run against to at least test if our happy path makes sense. I am worried that we are flying blind.

noot commented on crates/astria-conductor/src/executor.rs:
```diff
@@ -238,38 +238,62 @@ impl Executor {
         );
         self.sequencer_hash_to_execution_hash
-            .insert(block.block_hash, response.block_hash);
+            .insert(block.block_hash, response.block_hash.clone());
```
added a check at the top of this function to check if the block was already executed (ie. if `sequencer_hash_to_execution_hash` already contains an execution hash for this block hash) and returns it if it exists

noot commented on crates/astria-conductor/src/executor.rs:
```diff
@@ -238,38 +238,62 @@ impl Executor {
         Ok(())
     }
+
+    async fn finalize_block(
```
done!

noot commented:
@SuperFluffy addressed your comments and added unit tests for `execute_block` and `handle_block_received_from_data_availability` using a mock `ExecutionClient`. let me know if there's anything else you'd like me to add tests for!

SuperFluffy reviewed (commented):
I left a few nits that I would like to see addressed but approving this PR
